### PR TITLE
feat(huff_assert) Stack traces

### PR DESCRIPTION
**Pull request by iFrostizz**
## Overview

This PR aims to implement stack assertions. See https://github.com/huff-language/huff-rs/issues/132
When writing huff code, it's frequent to have a revert and is sometimes hard to debug. These assertions should makes us more confident on how the stack looks like at each line, and makes debugging easier.

After annotating the assertions, this is what `SAFE_ADD` from [SafeMath](https://github.com/pentagonxyz/huffmate/blob/main/src/math/SafeMath.huff) would look like:

```
#define macro SAFE_ADD() = takes (2) returns (1) {
    dup2                    $ [num2, num1, num2]
    add                     $ [result, num2]
    dup1                    $ [result, result, num2]
    swap2                   $ [num2, result, result]
    gt                      $ [is_overflow, result]
    iszero                  $ [is_not_overflow, result]
    is_not_overflow jumpi   $ [result]
        0x00 0x00 revert
    is_not_overflow:        $ [result]
}
```

Running:

```sh
huffc --check-assert src/SafeMath.huff
```

shouldn't return anything.

Now what happens if assertions are wrong ?

```
#define macro SAFE_ADD() = takes (2) returns (1) {
    dup2                    $ [num2, num1, num2, nope]
    add                     $ [result, num2]
    dup1                    $ [result, result, num2, definitely_not]
    swap2                   $ [num2, result, result]
    gt                      $ [is_overflow, result]
    iszero                  $ [is_not_overflow, result]
    is_not_overflow jumpi   $ [result]
        0x00 0x00 revert
    is_not_overflow:        $ [result]
}
```

it's gonna give
image

Some things left to-do:
- Custom stack & calldata
At the moment, there is no calldata in the call that is done to the macro, and stack is pushed with "0" bytes considering the `takes` parameter. Because of this, assertion checks are failing in some cases:
    - MAIN macro has no revert case if no function selector is matched
    - Pushing 0 could make the macro revert if dividing by 0 in the `SAFE_DIV` or `SAFE_MOD` macro

Not sure if we should do it in the command, like: 

```sh
huffc --check-assert src/SafeMath.huff <MACRO> <CALLDATA> <STACK>
```

and maybe make `check-assert` a subcommand then

## Checklist

- [x] add `--check-assert` command
- [x] Support input stacks and custom calldata
- [x] Support comments in assertions
- [x] Enforce `returns` count
- [x] Support variable instanciation
- [x] Display error lines
- [x] Write tests

### CODE REVIEW

**Comment by iFrostizz** [comment]
Custom input stack and calldata should be pretty much done. The "best" workaround I used was a `Vec<String>` for each field in the `check-assert` subcommand. So let's take SafeMath.huff, the command used to run the inspection would be:
```sh
huffc -s ~/CLionProjects/huff-testing/src/SafeMath.huff check-assert -m \ 
SAFE_DIV SAFE_DIV_WRAPPER SAFE_MOD MAIN \
 -d x 123456780000000000000000000000000000000a00000000000000000000000000000002 \
  x e6cb9013  -s 0000000000000000000000000000000a00000000000000000000000000000002 x \
0000000000000000000000000000000a00000000000000000000000000000002 x
```

Reference:

- x stands for "empty". Available for both stack and data
- _ stands for "autofill". Only working for stack, and is going to fill as much zeros as the number in "takes"
- if no custom stack is set, then it's gonna autofill by default

For some reason in SafeMath.huff the inspection is reverting for both `SAFE_MOD_WRAPPER` and `SAFE_DIV_WRAPPER`.
It seems that the stack that is passed between macros is set to 0 and the inspection (check for assertions) is not even ran.

The command is pretty verbose. Might be worth storing them in bash files to relax the terminal a bit

Any thoughts ?

**Comment by devtooligan** [comment]
@iFrostizz Hey fren.  Trying to give some love to this older pr.  Whats the status? Can it be reviewed? Looks like there are some merge conflicts now

**Comment by iFrostizz** [comment]
Hey @devtooligan I'd like to give it a self-review first once I have more bandwidth because I might be able to catch stuff. If anyone want to do it though, feel free.
Imo, the command-line syntax is too verbose and problematic.
Also, we are having issues with stack-changing macros because of loops.

File: huff_assert/tests/lexer.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        .into_iter()
+        .map(|x| x.unwrap())
+        .filter(|x| matches!(x.kind, TokenKind::Stack(_)))
+        .collect::<Vec<Token>>();
+
+    assert_eq!(tokens.len(), 4);
+    assert_eq!(tokens.get(0).unwrap().kind, TokenKind::Stack("$ [val]".to_string()));
+    assert_eq!(
+        tokens.get(1).unwrap().kind,
+        TokenKind::Stack("$ [/*oth_val, */ oth_val, val]".to_string())
```
Comment (iFrostizz) [inline huff_assert/tests/lexer.rs:43]: This is not a good idea, let's just remove the commented stack annotations for the lexing

File: huff_assert/tests/lexer.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    let flattened_source = FullFileSource { source, file: None, spans: vec![] };
+    let lexer = Lexer::new(flattened_source);
+    let tokens = lexer
+        .into_iter()
+        .map(|x| x.unwrap())
+        .filter(|x| matches!(x.kind, TokenKind::Stack(_)))
+        .collect::<Vec<Token>>();
+
+    assert_eq!(tokens.len(), 4);
+    assert_eq!(tokens.get(0).unwrap().kind, TokenKind::Stack("$ [val]".to_string()));
```
Comment (iFrostizz) [inline huff_assert/tests/lexer.rs:66]: For the data structure, that could be `TokenKind::Stack<Vec<TokenKind::StackElement>>` where `StackElement` contains the assertion name, to potentially support more features

File: huff_assert/tests/runner.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        vec![],
+    );
+
+    assert_eq!(res.name, "TEST");
+    assert_eq!(res.errors.len(), 1);
+
+    let err = res.errors.get(0).unwrap();
+    assert_eq!(err.kind, ErrorKind::Returns);
+    assert_eq!(err.expected, "`returns(0)`");
+    assert_eq!(err.got, "`[16]`");
```
Comment (iFrostizz) [inline huff_assert/tests/runner.rs:150]: Should not use a string slice here, use the token representation or a string only surrounding `16`

File: huff_assert/tests/runner.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        contract.macros.get(0).unwrap(),
+        String::default(),
+        U256::zero(),
+        vec![],
+    );
+
+    assert_eq!(res.name, "TEST");
+    assert_eq!(res.errors.len(), 0);
+
+    assert_eq!(res.reason, Return::Revert); // didn't reverted
```
Comment (iFrostizz) [inline huff_assert/tests/runner.rs:179]: weird comment

File: huff_cli/src/huffc.rs [hunk 5]
```diff
@@ -83,14 +87,15 @@ struct Huff {
...
-    /// Test subcommand
     #[clap(subcommand)]
-    test: Option<TestCommands>,
+    sub: Option<Subcommands>,
 }
 
 #[derive(Subcommand, Clone, Debug)]
-enum TestCommands {
-    /// Test subcommand
+enum Subcommands {
```
Comment (iFrostizz) [inline huff_cli/src/huffc.rs:95]: As this is now a subcommand, the current `huffc some/path/evm.huff` isn't working anymore. Any workaround ?

File: huff_cli/src/huffc.rs [hunk 6]
```diff
@@ -190,33 +216,72 @@ fn main() {
...
-                let match_ = Rc::new(match_);
-
-                for contract in &contracts {
-                    let tester = HuffTester::new(contract, Rc::clone(&match_));
-
-                    let start = Instant::now();
-                    match tester.execute() {
-                        Ok(res) => {
-                            print_test_report(res, ReportKind::from(&format), start);
+    if let Some(sub) = cli.sub {
```
Comment (iFrostizz) [inline huff_cli/src/huffc.rs:219]: move all this complexity in other functions, with better error handling using eyre ?

File: huff_codegen/src/lib.rs [hunk 7]
```diff
@@ -363,41 +395,46 @@ impl Codegen {
...
-                            if !&formatted_bytes.0[jump.bytecode_index + 2..jump.bytecode_index + 6]
-                                .eq("xxxx")
-                            {
-                                tracing::error!(
-                                    target: "codegen",
-                                    "JUMP DESTINATION PLACEHOLDER NOT FOUND FOR JUMPLABEL {}",
-                                    jump.label
-                                );
-                            }
+                if !formatted_bytes.0.starts_with("stack: ") {
```
Comment (iFrostizz) [inline huff_codegen/src/lib.rs:398]: this is dirty af

**Pull request closed without merging**